File: cpp_100-999/300.cc
// C++
// leetcode 300
// https://leetcode-cn.com/problems/longest-increasing-subsequence/
// dp[i]用来存储以nums[i]为结尾的最长子序列的大小
// 遍历求一遍dp[0] ~ dp[size-1]
// max(dp[0], ... , dp[size-1])就是答案
 
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
    	int size =nums.size(), ret = 1, temp = 0;
    	int dp[size];
    	dp[0] = 1;
    	for(int i = 1; i<size; i++){
    		temp = nums[i];
    		dp[i] = 1;
    		for(int j =0; j<i; j++){
    			if(temp>nums[j] && dp[j]>=dp[i]){
    				dp[i] = dp[j]+1;
    			}
    		}
    		if(dp[i]>ret)
    			ret =dp[i];
    	}
    	return ret;
    }
};

File: Go/Go_100-999/300.go
// Go
// leetcode 300
// https://leetcode.cn/problems/longest-increasing-subsequence/

func lengthOfLIS(nums []int) int {
	n := len(nums)
	smallestValForLength := make([]int, n)
	for i := 0; i < n; i++ {
		smallestValForLength[i] = 20000
	}
	for i, x := range nums {
		for j := i-1; j >= 0; j-- {
			if x > smallestValForLength[j] {
				smallestValForLength[j+1] = min(smallestValForLength[j+1], x)
			}
		}
        smallestValForLength[0] = min(smallestValForLength[0], x)
	}
    // fmt.Println(smallestValForLength)
	for i := n-1; i >= 0; i-- {
		if smallestValForLength[i] < 20000 {
			return i+1
		}
	}
	return 0
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}